### ISSUE TITLE
Speed downgrade with increasing number of SNodeTree 

### ISSUE THREAD

**Issue by ifsheldon** [taichi-dev/taichi#5282]
**Describe the bug**
The speed of the creation of a `SNodeTree` slows down linearly as the number of un-destroyed/existing `SNodeTree` increases, which leads to a time complexity of O(N^2) where N is the total number of created `SNodeTree`.

**To Reproduce**

Run the code below and plot iterations/second, you will see the speed slows down linearly as existing `SNodeTree` increases.

```python
import taichi as ti
from tqdm import tqdm

if __name__ == "__main__":
    ti.init(ti.cpu)
    iterations = 2
    snode_tree_num = 1000 # Or 512 without relaxing `constexpr int kMaxNumSnodeTreesLlvm` in taichi/inc/constants.h
    for i in range(iterations):
        snode_trees = []
        for j in tqdm(range(snode_tree_num)):
            x = ti.field(ti.f32)
            fb = ti.FieldsBuilder()
            fb.place(x)
            snodetree = fb.finalize()
            snode_trees.append(snodetree)
        for snodetree in snode_trees:
            snodetree.destroy()
```

**Additional comments**
In my use case, I potentially need to create a lot of `SNodeTree`s at runtime and thus I self-compiled Taichi with a larger `kMaxNumSnodeTreesLlvm` (10K) but this O(N^2) time complexity is not affordable for me.

**Comment by lin-hitonami** [comment]
Solved in #5511

**Comment by ifsheldon** [comment]
Mega thanks to @lin-hitonami !
